File: python/scenes/outdoor_env.py
#!/usr/bin/env python3
"""Outdoor environment scene for LumenCore."""
from __future__ import annotations

import os
import sys
from pathlib import Path

import lumencore as lc


def resolve_asset(relative: str) -> str:
    candidates = []
    root = os.environ.get("LUMENCORE_ROOT")
    if root:
        candidates.append(str(Path(root) / relative))
    candidates.extend(
        [
            relative,
            str(Path("..") / relative),
            str(Path("../..") / relative),
            str(Path("/work") / relative),
        ]
    )
    for path in candidates:
        if Path(path).is_file():
            return path
    return candidates[0]


def main() -> int:
    out = sys.argv[1] if len(sys.argv) > 1 else "outdoor_env.png"
    spp = int(sys.argv[2]) if len(sys.argv) > 2 else 256
    denoise = (int(sys.argv[3]) != 0) if len(sys.argv) > 3 else True

    scene = lc.Scene()
    scene.load_env_map(resolve_asset("assets/env/studio.hdr"))
    grass = scene.add_material(lc.Material(base_color=(0.2, 0.35, 0.15), roughness=0.95))
    stone = scene.add_material(lc.Material(base_color=(0.45, 0.45, 0.48), roughness=0.7))
    chrome = scene.add_material(lc.Material(base_color=(0.95, 0.95, 0.98), metallic=1.0, roughness=0.02))
    glass = scene.add_material(lc.Material(base_color=(1, 1, 1), roughness=0.0, transmission=1.0, ior=1.5))
    warm = scene.add_material(lc.Material(base_color=(0.85, 0.45, 0.2), roughness=0.4))

    scene.add_mesh(lc.make_quad((-20, 0, -20), (40, 0, 0), (0, 0, 40), grass))
    scene.add_mesh(lc.make_box((-1.2, 0, -0.4), (-0.3, 1.6, 0.5), stone))
    scene.add_mesh(lc.make_uv_sphere((0.8, 0.6, 0.2), 0.6, chrome))
    scene.add_mesh(lc.make_uv_sphere((-0.1, 0.4, 1.4), 0.4, glass))
    scene.add_mesh(lc.make_uv_sphere((1.8, 0.35, 1.2), 0.35, warm))

    # Soft fill; HDRI carries most of the lighting / reflections
    light_corner = (4.0, 8.0, -2.0)
    light_u = (2.5, 0, 0.5)
    light_v = (0, 0, 2.5)
    light_mat = scene.add_material(lc.Material(base_color=(0, 0, 0), roughness=1.0, emission=(12, 11, 9)))
    scene.add_mesh(lc.make_quad(light_corner, light_u, light_v, light_mat))
    scene.add_quad_light(light_corner, light_u, light_v, (12, 11, 9))

    scene.background_top = (0.08, 0.1, 0.14)
    scene.background_bottom = (0.04, 0.04, 0.05)

    camera = lc.Camera(
        eye=(-2.8, 1.8, 4.2),
        lookat=(0.3, 0.5, 0.4),
        fov_y_deg=40,
        aspect=16 / 9,
        aperture=0.04,
        focus_dist=4.5,
    )
    cfg = lc.RenderConfig(width=2560, height=1440, spp=spp, denoise=denoise, output_path=out)
    lc.Renderer().render(scene, camera, cfg)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/outdoor_env.cc
#include "nrtx/nrtx.h"

#include <cstdlib>
#include <iostream>
#include <string>

using namespace nrtx;

int main(int argc, char **argv) {
  const std::string out = argc > 1 ? argv[1] : "outdoor_env.png";
  const int spp = argc > 2 ? std::atoi(argv[2]) : 256;
  const bool denoise = argc > 3 ? std::atoi(argv[3]) != 0 : true;

  Scene scene;
  const int grass = scene.add_material({make_float3(0.2f, 0.35f, 0.15f), 0, 0.95f, 0, 1.5f, {}});
  const int stone = scene.add_material({make_float3(0.45f, 0.45f, 0.48f), 0, 0.7f, 0, 1.5f, {}});
  const int chrome = scene.add_material({make_float3(0.95f, 0.95f, 0.98f), 1.0f, 0.02f, 0, 1.5f, {}});
  const int glass = scene.add_material({make_float3(1.0f, 1.0f, 1.0f), 0, 0.0f, 1.0f, 1.5f, {}});
  const int warm = scene.add_material({make_float3(0.85f, 0.45f, 0.2f), 0, 0.4f, 0, 1.5f, {}});

  scene.add_mesh(make_quad(make_float3(-20, 0, -20), make_float3(40, 0, 0), make_float3(0, 0, 40), grass));
  scene.add_mesh(make_box(make_float3(-1.2f, 0, -0.4f), make_float3(-0.3f, 1.6f, 0.5f), stone));
  scene.add_mesh(make_uv_sphere(make_float3(0.8f, 0.6f, 0.2f), 0.6f, chrome));
  scene.add_mesh(make_uv_sphere(make_float3(-0.1f, 0.4f, 1.4f), 0.4f, glass));
  scene.add_mesh(make_uv_sphere(make_float3(1.8f, 0.35f, 1.2f), 0.35f, warm));

  // Soft sun-like distant area light
  const float3 light_corner = make_float3(4.0f, 8.0f, -2.0f);
  const float3 light_u = make_float3(2.5f, 0, 0.5f);
  const float3 light_v = make_float3(0, 0, 2.5f);
  const int light_mat =
      scene.add_material({make_float3(0, 0, 0), 0, 1, 0, 1.5f, make_float3(40.0f, 36.0f, 28.0f)});
  scene.add_mesh(make_quad(light_corner, light_u, light_v, light_mat));
  scene.add_quad_light(light_corner, light_u, light_v, make_float3(40.0f, 36.0f, 28.0f));

  scene.background_top = make_float3(0.45f, 0.65f, 0.95f);
  scene.background_bottom = make_float3(0.75f, 0.8f, 0.85f);

  Camera camera;
  camera.eye = make_float3(-2.8f, 1.8f, 4.2f);
  camera.lookat = make_float3(0.3f, 0.5f, 0.4f);
  camera.fov_y_deg = 40.0f;
  camera.aspect = 16.0f / 9.0f;
  camera.aperture = 0.04f;
  camera.focus_dist = 4.5f;

  RenderConfig cfg;
  cfg.width = 2560;
  cfg.height = 1440;
  cfg.spp = spp;
  cfg.denoise = denoise;
  cfg.output_path = out;

  try {
    Renderer renderer;
    renderer.render(scene, camera, cfg);
  } catch (const std::exception &ex) {
    std::cerr << "Error: " << ex.what() << "\n";
    return 1;
  }
  return 0;
}
